test/dynamo/test_config.py

```diff
@@ -110,6 +110,206 @@ def test_config_hash(self):
+            )(fn)
+            opt_fn_static_shape(torch.randn(2), torch.randn(2))
+            opt_fn_static_shape(torch.randn(3), torch.randn(3))
+
+        self.assertEqual(cnt_dynamic.frame_count, 2)
+
+        with torch._dynamo.config.patch(
+            automatic_dynamic_shapes=True, assume_static_by_default=False
+        ):
+            for i in range(2, 12):
+                opt_fn_static_shape(
+                    torch.randn(i), torch.randn(i)
+                )  # will be recompiled under new config
+
+        self.assertEqual(cnt_dynamic.frame_count, 3)
+
+    @disable_cache_limit()
+    def test_no_saved_config_nested(self):
+        def fn(a, b):
+            return a - b * 10
+
+        torch._dynamo.reset()
+        cnt_dynamic = torch._dynamo.testing.CompileCounter()
+        cnt_dynamic_1 = torch._dynamo.testing.CompileCounter()
+        with torch._dynamo.config.patch(
+            automatic_dynamic_shapes=True, assume_static_by_default=False
+        ):
+            opt_fn_static_shape = torch._dynamo.optimize(cnt_dynamic, dynamic=False)(fn)
+
+            # Will trigger recompile as compiled as static
+            opt_fn_static_shape(torch.randn(2), torch.randn(2))
+            opt_fn_static_shape(torch.randn(3), torch.randn(3))
+
+            self.assertEqual(cnt_dynamic.frame_count, 2)
+
+            opt_fn_try_dynamic = torch._dynamo.optimize(
+                cnt_dynamic_1, save_config=False
+            )(opt_fn_static_shape)
+
+            for i in range(2, 6):
+                opt_fn_try_dynamic(torch.randn(i), torch.randn(i))
+            self.assertEqual(cnt_dynamic_1.frame_count, 1)
+
+        # Saved config = False will use whatever config is available
+        with torch._dynamo.config.patch(
+            automatic_dynamic_shapes=False, assume_static_by_default=True
+        ):
+            for i in range(6, 12):
+                opt_fn_try_dynamic(torch.randn(i), torch.randn(i))
+            self.assertEqual(cnt_dynamic_1.frame_count, 7)
+
+    @disable_cache_limit()
+    def test_config_changed_from_guarded_config_1(self):
+        def fn(a, b):
+            return a - b * 10
+
+        torch._dynamo.reset()
+
+        cnt_dynamic = torch._dynamo.testing.CompileCounter()
+        with torch._dynamo.config.patch(
+            automatic_dynamic_shapes=False, assume_static_by_default=True
+        ):
+            opt_fn_static_shape = torch._dynamo.optimize(cnt_dynamic)(fn)
+            res = opt_fn_static_shape(torch.randn(2), torch.randn(2))
+            print("RES", res)
+            opt_fn_static_shape(torch.randn(3), torch.randn(3))
+
+        self.assertEqual(cnt_dynamic.frame_count, 2)
+
+        with torch._dynamo.config.patch(
+            automatic_dynamic_shapes=True, assume_static_by_default=False
+        ):
+            for i in range(2, 12):
+                # Only 4-11 will now be recompiled under old config
+                # 2-3 have been recompiled under old config due to shape mismatch
```

Comment:
I don't understand this comment, but the test case looks right. Shouldn't 2-3 cache hit from the previous run? The important thing is we DO recompile even though dynamic was requested, because we're using the old config here.

Reply:
Yes, I think it means 2-3 have already been (re)compiled. Let just say "should hit cache"